# Adds libnatspec support
# Vitaly Lipatov <lav@etersoft.ru>

--- /home/lav/work/build/BUILD/submount-0.9/submountd-0.9/configure.ac	2005-02-26 16:08:47 +0300
+++ submountd-0.9/configure.ac	2005-02-26 16:00:03 +0300
@@ -37,6 +37,9 @@
 fi
 AC_SUBST(LIB_RESMGR)
 
+# Checks for libnatspec
+AM_PATH_NATSPEC
+				
 AC_PREFIX_DEFAULT(/)
 AC_CONFIG_FILES([Makefile])
 AC_OUTPUT

--- /home/lav/work/build/BUILD/submount-0.9/submountd-0.9/submountd.c	2005-02-26 16:08:47 +0300
+++ submountd-0.9/submountd.c	2005-03-01 17:28:37 +0300
@@ -40,6 +40,10 @@
 #include <pwd.h>
 #endif
 
+#ifdef HAVE_NATSPEC
+#include <natspec.h>
+#endif
+
 #include "submountd.h"
 
 
@@ -190,6 +194,10 @@
 		}
 	}
 #endif
+
+#ifdef HAVE_NATSPEC
+	options = natspec_get_enriched_fs_options(fstype, options);
+#endif
 	retval = mount(device, mountpoint, fstype, flags, options);
 
 	if (retval) {
@@ -198,9 +206,11 @@
 			retval = do_mount(device, mountpoint, fstype, flags,
 				  options);
 		}
-		return retval;
 	}
-	return 0;
+#ifdef HAVE_NATSPEC
+	free(options);
+#endif
+	return retval;
 }
 
 

--- /home/lav/work/build/BUILD/submount-0.9/submountd-0.9/Makefile.am	2005-02-20 17:24:46 +0300
+++ submountd-0.9/Makefile.am	2005-02-20 15:49:35 +0300
@@ -2,7 +2,7 @@
 sbin_PROGRAMS = submountd net-submountd
 
 submountd_SOURCES = submountd.c mount_guess_fstype.c
-submountd_LDADD = $(LIB_RESMGR)
+submountd_LDADD = $(LIB_RESMGR) $(NATSPEC_LIBS)
 net_submountd_SOURCES = net-submountd.c
 man_MANS = submount.8
 
